llvm/lib/CAS/Utils.cpp

```diff
@@ -84,3 +84,34 @@ Error cas::walkFileTreeRecursively(
 
   return Error::success();
 }
+
+Error cas::walkObjectsRecursively(
+    CASDB &CAS, CASID ID,
+    function_ref<Error(const std::pair<CASID, unsigned> &)> Callback) {
+
+  BumpPtrAllocator Alloc;
+  StringSaver Saver(Alloc);
+  SmallString<128> PathStorage;
+  SmallVector<std::pair<CASID, unsigned>> Stack;
+  Stack.emplace_back(ID, 0);
+
+  while (!Stack.empty()) {
+    if (Error E = Callback(Stack.back()))
+      return E;
+
+    std::pair<CASID, unsigned> Parent = Stack.pop_back_val();
+    Expected<NodeProxy> ExpNode = CAS.getNode(Parent.first);
+    if (Error E = ExpNode.takeError())
+      return E;
+    NodeProxy Node = *ExpNode;
+    for (int I = Node.getNumReferences(), E = 0; I != E; --I) {
+      Optional<CASID> Child = Node.getReference(I - 1);
+      if (!Child)
```

Comment:
When does this condition hit? I don't see the reason why getReference should return optional?

Reply:
This condition hits when getReference returns None, because:

```
Optional<CASID> NodeProxy::getReference(size_t I) const {
  return CAS->getReferenceInNode(*this, I);
}
```
Which calls:
```
Optional<CASID> OnDiskCAS::getReferenceInNode(const NodeProxy &Node,
                                              size_t I) const {
  DataRecordHandle Record = DataRecordHandle::get(
      reinterpret_cast<char *>(const_cast<void *>(getNodePtr(Node))));

  Optional<CASID> ID = getCASID(Record.getRefs()[I]);
  return ID;
}
```
Which calls:
```
Optional<CASID> OnDiskCAS::getCASID(InternalRef Ref) const {
  switch (Ref.getOffsetKind()) {
  case InternalRef::OffsetKind::String2B:
    /// Strings are not exposed.
    return None;

  case InternalRef::OffsetKind::DataRecord: {
    DataRecordHandle Handle =
        DataRecordHandle::get(DataPool.beginData(Ref.getFileOffset()));
    return getIDFromIndexOffset(Handle.getTrieRecordOffset());
  }

  case InternalRef::OffsetKind::IndexRecord:
    return getIDFromIndexOffset(Ref.getFileOffset());
  }
}
```

And atleast in my contrived test case, there were times when the `InternalRef::OffsetKind` was `String2B` which means that `getCASID` returned None, which means that `OnDiskCAS::getReferenceInNode` would have asserted and crashed. Please see changes in OnDiskCAS.cpp as well

Reply:
That seems to be the wrong reason. Having a String2B node in the references doesn't mean it can be skipped. It needs to be visited too as a special inline node. This is the unnecessary complication from optimization.

Reply:
So I am not sure what to do here, when I try to get a `String2B` node, if I leave `OnDiskCAS::getReferenceInNode` as it is, it will hit an assert. How do you think a `String2B` node should be handled?

Reply:
I am not sure you have the best option here. I don't think it is most desirable to skip those blocks or exposing InternalRef to the visitor, or just get a Optional<CASID> for callback.

Maybe you can find something better after Duncan finishes his rewrite here: https://github.com/apple/llvm-project/pull/4230